### PR TITLE
Expanded tests for germline-assembly caller

### PR DESCRIPTION
**Pull request by arahuja**
This expands the test suite for the germline assembly caller, adding specific variants to find of different varieties.

The test suite is also altered to create a single spark context for the full suite (as opposed to a spark context per test). This lets us load the reads/reference once rather than in each test.

This also fixes an issue with merging nodes in the DeBruijn graph (nodes were not being properly removed), a test case has been added.

Additionally, makes two changes to the germline caller itself:
- Instead of search from a reference from start of the first read, it uses the start with the window. The first read may extend earlier than the start of the window, and is likely the only read that stretches that far back. Therefore, only paths supported by that _single_ read would have been found.
- Instead of first doing a pass to find variant pileups, this checks this in the windows as it is processing. It find's variable windows, and only performs the assembly on those regions.

[Review on Reviewable](https://reviewable.io/reviews/hammerlab/guacamole/382)

### CODE REVIEW

File: src/main/scala/org/hammerlab/guacamole/assembly/DeBruijnGraph.scala [hunk 1]
```diff
@@ -212,9 +223,16 @@ class DeBruijnGraph(val kmerSize: Int,
     while (frontier.nonEmpty && paths.size < maxPaths) {
       val next = frontier.pop()
 
+      if (debugPrint) {
+        if (currentPath.length == 0) {
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/assembly/DeBruijnGraph.scala]: `.isEmpty`

File: src/main/scala/org/hammerlab/guacamole/alignment/AffineGapPenaltyAlignment.scala [hunk 2]
```diff
@@ -37,12 +37,12 @@ object AffineGapPenaltyAlignment {
         closeGapProbability
       )
 
-    val ((refStartIdx, path, score), refEndIdx) =
+    val ((refEndIdx, path, score), refStartIdx) =
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/alignment/AffineGapPenaltyAlignment.scala]: was this mix-up causing broken behavior?

File: src/main/scala/org/hammerlab/guacamole/assembly/DeBruijnGraph.scala [hunk 3]
```diff
@@ -241,13 +260,18 @@ class DeBruijnGraph(val kmerSize: Int,
...
           // Found legitimate path to sink, save path
-          paths = (currentPath.reverse) :: paths
+          paths = currentPath.reverse :: paths
         } // else degenerate path, too long or too short
 
         if (lastBranchIndex.nonEmpty) {
           // Backtrack the current path to the last node with siblings or start
           val backtrackIndex = lastBranchIndex.top
+          // Remove the nodes from the last branch from the visited set
+          currentPath.slice(0, currentPath.length - backtrackIndex).map(visited.remove)
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/assembly/DeBruijnGraph.scala]: nit: `s/map/foreach/`?

File: src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala [hunk 4]
```diff
@@ -104,12 +104,18 @@ object GermlineAssemblyCaller {
       val referenceKmerSink = currentReference.takeRight(kmerSize)
 
       // If the current window size doesn't cover the kmer size we won't be able to find the reference start/end
-      if (currentReference.size < kmerSize)
+      if (currentReference.size < kmerSize || referenceKmerSource == referenceKmerSink)
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala]: hm, I'm not sold on this added clause. is there nothing useful we can do in regions where the reference as a k-repeat? I can imagine a reasonable-sized window with identical `k`-mers on each end where we'd still want to call variants inside it..

File: src/main/scala/org/hammerlab/guacamole/alignment/ReadAlignment.scala [hunk 5]
```diff
@@ -32,7 +32,8 @@ object AlignmentState extends Enumeration {
  * @param alignmentScore Score of the alignment
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/alignment/ReadAlignment.scala]: add the new params? they're a closed interval?

File: src/test/scala/org/hammerlab/guacamole/assembly/DeBrujinGraphSuite.scala [hunk 6]
```diff
@@ -229,6 +229,32 @@ class DeBruijnGraphSuite extends GuacFunSuite with Matchers {
...
+      Seq(longRead),
+      kmerSize,
+      minOccurrence = 1,
+      mergeNodes = true
+    )
+
+    longRead
+      .sliding(kmerSize)
+      .zipWithIndex.map(
+        {
```
Comment (ryan-williams) [inline src/test/scala/org/hammerlab/guacamole/assembly/DeBrujinGraphSuite.scala]: nit: IJ doesn't like the parends around the braces here

Comment (ryan-williams) [inline src/test/scala/org/hammerlab/guacamole/assembly/DeBrujinGraphSuite.scala]: also wants `s/map/foreach/`

File: src/main/scala/org/hammerlab/guacamole/alignment/AffineGapPenaltyAlignment.scala (hunk 2 above)
Comment (arahuja) [inline src/main/scala/org/hammerlab/guacamole/alignment/AffineGapPenaltyAlignment.scala]: Somewhat, they weren't used anywhere, but when I added some tests and wanted to use them I noticed.

File: src/test/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCallerSuite.scala [hunk 7]
```diff
@@ -1,34 +1,203 @@
...
+  args.parallelism = 2
 
-    GermlineAssemblyCaller.Caller.run(args, sc)
+  var sc: SparkContext = _
+  var reference: ReferenceBroadcast = _
+  var readSet: ReadSet = _
 
-    val variants = new ADAMContext(sc).loadVariants(args.variantOutput).collect()
+  before {
+    sc = Common.createSparkContext()
```
Comment (ryan-williams) [inline src/test/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCallerSuite.scala]: I was struggling with this recently; surprised that `sparkTest`, `sparkBefore` etc. each set up and tear down a SparkContext! Will have to use this pattern more often

File: src/main/scala/org/hammerlab/guacamole/alignment/ReadAlignment.scala [hunk 8]
```diff
@@ -32,7 +32,8 @@ object AlignmentState extends Enumeration {
...
  */
 case class ReadAlignment(alignments: Seq[AlignmentState],
-                         refBases: Seq[Byte],
```
Comment (arahuja) [inline src/main/scala/org/hammerlab/guacamole/alignment/ReadAlignment.scala]: Sorry, didn't mean you delete your comment here @ryan-williams , meant to delete my empty response

File: src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala (hunk 4 above)
Comment (arahuja) [inline src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala]: Yes, definitely possible, but as this is setup, I don't see a good way to handle it for now. Possible solutions may be just increasing the kmer size and looking again, but just punting on it for now.

File: src/test/scala/org/hammerlab/guacamole/assembly/DeBrujinGraphSuite.scala (hunk 6 above)
Comment (arahuja) [inline src/test/scala/org/hammerlab/guacamole/assembly/DeBrujinGraphSuite.scala]: > nit: IJ doesn't like the parends around the braces here

weird, that looks weird to me, but I'm ok with it

File: src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala (hunk 4 above)
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/commands/GermlineAssemblyCaller.scala]: OK, punting wfm. I see that there is a more general inability to handle regions with repeating k-mers atm. I wonder if that happens frequently? Adding a TODO and/or a warning about this case might make sense.

Basically any time we skip a node in the DBG DFS due to it already being in `visited`, we are discarding a possible valid path with a repeat segment in it, right?

**Comment by ryan-williams** [comment]
lgtm % one comment about whether we should leave any breadcrumbs in the repeat-kmer case

**Comment by arahuja** [comment]
@ryan-williams thanks for reviewing! Added a warning on the invalid start and end kmers, should be useful for diagnosing when/how often it occurs.

**Pull request merged**